### PR TITLE
[SPARK-30019][SQL] Add the owner property to v2 table

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

Add `owner` property to v2 table, it is reversed by `TableCatalog`, indicates the table's owner.

### Why are the changes needed?

enhance ownership management of catalog API

### Does this PR introduce any user-facing change?

yes, add 1 reserved property - `owner` , and it is not allowed to use in OPTIONS/TBLPROPERTIES anymore, only if legacy on

### How was this patch tested?

add uts

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala [hunk 1]
```diff
@@ -227,13 +228,14 @@ case class AtomicReplaceTableAsSelectExec(
 
   override protected def doExecute(): RDD[InternalRow] = {
     val schema = query.schema.asNullable
+    val propertiesWithOwner = withDefaultOwnership(properties).asJava
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala]: can we add an analyzer rule to catch all the CREATE TABLE like commands and add the ownership properties?

Comment (yaooqinn) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala]: do we have any other APIs will bypass the analyzer?

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala]: If we only look at the logical plan, I think only the following commands need to be considered:
1. CREATE TABLE
2. CREATE TABLE AS SELECT
3. REPLACE TABLE
4. REPLACE TABLE AS SELECT

Comment (yaooqinn) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala]: Yes, they have a common parent `V2CreateTablePlan` when they are on the logical side. We may add a method like 
```scala
override def withPartitioning(rewritten: Seq[Transform]): V2CreateTablePlan = {
    this.copy(partitioning = rewritten)
}

override def withOwnership(): V2CreateTablePlan = {
    this.copy(properties = this.properites ++ ownership)
}
```

but not more elegant and simpler as it is now

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala]: a logical plan may have multiple physical plans, e.g. the atomic version of CTAS. That's why I think it's better to do it at the logical phase.

Comment (yaooqinn) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala]: maybe we can add them in DataSourceV2Strategy? then we don't need extra analyzer rules to copy themselves

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala]: that also works

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 2]
```diff
@@ -569,7 +570,11 @@ private[hive] class HiveClientImpl(
 
   override def createTable(table: CatalogTable, ignoreIfExists: Boolean): Unit = withHiveState {
     verifyColumnDataType(table.dataSchema)
-    client.createTable(toHiveTable(table, Some(userName)), ignoreIfExists)
+    val ownerType =
+      table.properties.getOrElse(TableCatalog.PROP_OWNER_TYPE, PrincipalType.USER.name())
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: when there is no owner type in table properties?

**Comment by cloud-fan** [comment]
shall we also print owner name/type in DESC TABLE like DESC NAMESPACE?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala (hunk 2 above)
Comment (yaooqinn) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: the hive catalog's ownership is decided by this class as Database does,  it's simpler. For v2 commands, we have to do injection in our plans because we are not the catalog developer. As if we are, we also only need to inject ownership in that catalog's `createTable `

**Comment by yaooqinn** [comment]
> shall we also print owner name/type in DESC TABLE like DESC NAMESPACE?

The `DescribeTableExec` will show all reserved properties separately 

https://github.com/apache/spark/blob/16da714ea5af9105a98e03d71321949d18590506/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DescribeTableExec.scala#L52-L56

**Comment by cloud-fan** [comment]
I mean, we should follow `DESC NAMESPACE` to implement it, e.g. display `Owner Name: abc` which looks nicer. Anyway we can do it in a followup.

**Comment by cloud-fan** [comment]
LGTM

**Comment by yaooqinn** [comment]
retest this please

**Comment by yaooqinn** [comment]
retest this please

File: docs/sql-migration-guide.md [hunk 3]
```diff
@@ -331,7 +331,21 @@ license: |
...
-            For databases, please use the ALTER DATABASE ... SET OWNER syntax to modify it.
+            For databases, please use the ALTER DATABASE ... SET OWNER syntax to modify it
+          </td>
+        </tr>
+        <tr>
+          <td>
+            owner
+          </td>
+          <td>
+            no
```
Comment (cloud-fan) [inline docs/sql-migration-guide.md]: it's reserved for both table and namespace, or do I miss something?

Comment (cloud-fan) [reply docs/sql-migration-guide.md]: nvm, we can update when revert ALTER NAMESPACE SET OWNER.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 4]
```diff
@@ -2680,6 +2680,10 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
         throw new ParseException(s"$PROP_LOCATION is a reserved table property, please use" +
           s" the LOCATION clause to specify it.", ctx)
       case (PROP_LOCATION, _) => false
+      case (PROP_OWNER, _) if !legacyOn =>
+        throw new ParseException(s"$PROP_OWNER is a reserved table property , please use" +
+          " ALTER TABLE ... SET OWNER ... to specify it.", ctx)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: it can't be changed after table is created.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala [hunk 5]
```diff
@@ -36,6 +36,7 @@ import org.apache.spark.sql.catalyst.expressions.{Attribute, AttributeReference}
 import org.apache.spark.sql.catalyst.plans.DescribeTableSchema
 import org.apache.spark.sql.catalyst.plans.logical._
 import org.apache.spark.sql.catalyst.util.{escapeSingleQuotedString, quoteIdentifier, CaseInsensitiveMap}
+import org.apache.spark.sql.connector.catalog.TableCatalog
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala]: unnecessary change

File: sql/core/src/test/scala/org/apache/spark/sql/ShowCreateTableSuite.scala [hunk 6]
```diff
@@ -19,6 +19,7 @@ package org.apache.spark.sql
 
 import org.apache.spark.sql.catalyst.TableIdentifier
 import org.apache.spark.sql.catalyst.catalog.CatalogTable
+import org.apache.spark.sql.connector.catalog.TableCatalog
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/ShowCreateTableSuite.scala]: unnecessary

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 7]
```diff
@@ -54,6 +54,7 @@ import org.apache.spark.sql.catalyst.catalog.CatalogTypes.TablePartitionSpec
 import org.apache.spark.sql.catalyst.expressions.Expression
 import org.apache.spark.sql.catalyst.parser.{CatalystSqlParser, ParseException}
 import org.apache.spark.sql.connector.catalog.SupportsNamespaces._
+import org.apache.spark.sql.connector.catalog.TableCatalog
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: ditto

**Comment by cloud-fan** [comment]
can we add some tests to make sure CREATE/ALTER TABLE fails if setting `owner` property?

 Also let's update the PR title. Now we are just adding an owner property to v2 table.

**Comment by yaooqinn** [comment]
> can we add some tests to make sure CREATE/ALTER TABLE fails if setting `owner` property?
> 
https://github.com/apache/spark/blob/9bb78eb82fd3eecd401c4a0175e02a0f6aff925d/sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala#L906-951

> Also let's update the PR title. Now we are just adding an owner property to v2 table.
OK

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala [hunk 8]
```diff
@@ -1403,11 +1406,12 @@ abstract class DDLSuite extends QueryTest with SQLTestUtils {
     createDatabase(catalog, "dbx")
     createTable(catalog, tableIdent, isDatasourceTable)
     def getProps: Map[String, String] = {
-      if (isUsingHiveMetastore) {
+      val props = if (isUsingHiveMetastore) {
         normalizeCatalogTable(catalog.getTableMetadata(tableIdent)).properties
       } else {
         catalog.getTableMetadata(tableIdent).properties
       }
+      props -- TableCatalog.RESERVED_PROPERTIES.asScala
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala]: do we need this change?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala [hunk 9]
```diff
@@ -418,13 +418,26 @@ class HiveCatalogedDDLSuite extends DDLSuite with TestHiveSingleton with BeforeA
...
   }
+
+  test("Table Ownership") {
+    val catalog = spark.sessionState.catalog
+    try {
+      sql(s"CREATE TABLE spark_30019(k int)")
+      assert(sql(s"DESCRIBE TABLE EXTENDED spark_30019").where("col_name='Owner'")
+        .collect().head.getString(1) === Utils.getCurrentUserName())
+      val e2 = intercept[ParseException](
+        sql(s"CREATE TABLE spark_30019_2 WITH TBLPROPERTIES('owner'='spark_30019')"))
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala]: do we need to test it? I think we already have CREATE/REPLACE/ALTER TABLE tests for reserved properties.

**Comment by cloud-fan** [comment]
LGTM except a few minor comments

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala (hunk 9 above)
Comment (yaooqinn) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala]: you mean this whole test or this line?

**Comment by cloud-fan** [comment]
thanks, merging to master!
